src/templates/stepper.html

```diff
@@ -180,59 +197,109 @@ <h4><span class="help float-right"> By clicking the <code>Previous State</code>
+                        </thead>
+                        <tbody>
+                            <tr>
+                                <th class="w-auto">Trace State</th>
+                                {% for ps in prefixstates %}
+                                    <td><pre>{{ ps.formattedTraceAssignment }}</pre></td>
+                                {% endfor %}
+                                {% for ps in cyclestates %}
+                                    <td class="cycle-state-col"><pre>{{ ps.formattedTraceAssignment }}</pre></td>
+                                {% endfor %}
+                            </tr>
+                            <tr>
+                                <th class="w-auto">
+                                    Formula
+                                    </span>
+                                </th>
+                                {% for ps in prefixstates %}
+                                    <td>
+                                        {{ ps.formulaAsHTML | safe }}
+                                    </td>
+                                {% endfor %}
+                                {% for ps in cyclestates %}
+                                    <td class="cycle-state-col">
+                                        {{ ps.formulaAsHTML | safe }}
+                                    </td>
+                                {% endfor %}
+                            </tr>
+                        </tbody>
+                    </table>
+                {% else %}
+                    <div class="alert alert-info">
+                        <p><strong>No formula loaded.</strong></p>
+                        <p>Please enter an LTL formula and trace to see the table view.</p>
+                    </div>
+                {% endif %}
+            </div>
+            <!-- ================================ -->
+            <!-- MATRIX VIEW - Binary matrix display of subformula satisfaction over time -->
+            <!-- ================================ -->
+            <div class="tab-pane fade" id="matrix-view" role="tabpanel">
+                <p>
+                This matrix view shows how each subformula's satisfaction evolves over time.
+                Each row represents a subformula, and each column represents a time step.
+                Values are shown as <strong class="matrix-satisfied">1</strong> (satisfied) or <strong class="matrix-unsatisfied">0</strong> (unsatisfied).
+                </p>
+                
+                {% if matrix_data.subformulae %}
+                <!-- Matrix view table -->
+                <table class="table table-bordered table-sm" id="matrixTable">
+                    <thead>
+                        <tr>
+                            <th class="w-auto">Subformula</th>
+                            {% if matrix_data.rows and matrix_data.rows|length > 0 %}
+                                {% set num_columns = matrix_data.rows[0].values|length %}
```

Comment:
nder(app, template, context)
  File "/venv/lib/python3.9/site-packages/flask/templating.py", line 131, in _render
    rv = template.render(context)
  File "/venv/lib/python3.9/site-packages/jinja2/environment.py", line 1295, in render
    self.environment.handle_exception()
  File "/venv/lib/python3.9/site-packages/jinja2/environment.py", line 942, in handle_exception
    raise rewrite_traceback_stack(source=source)
  File "/src/templates/stepper.html", line 277, in top-level template code
    {% set num_columns = matrix_data.rows[0].values|length %}
TypeError: object of type 'builtin_function_or_method' has no len()